pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/util/HelixSetupUtils.java

```diff
@@ -53,126 +55,217 @@
-        List<String> states = curStateModelDef.getStatesPriorityList();
-        if (states.contains(PinotHelixSegmentOnlineOfflineStateModelGenerator.CONSUMING_STATE)) {
-          LOGGER.info("State model {} already updated to contain CONSUMING state", segmentStateModelName);
-          return;
-        } else {
-          LOGGER.info("Updating {} to add states for low level consumers", segmentStateModelName);
-          StateModelDefinition newStateModelDef =
-              PinotHelixSegmentOnlineOfflineStateModelGenerator.generatePinotStateModelDefinition();
-          ZkClient zkClient = new ZkClient(zkPath);
-          zkClient.waitUntilConnected(CommonConstants.Helix.ZkClient.DEFAULT_CONNECT_TIMEOUT_SEC, TimeUnit.SECONDS);
-          zkClient.setZkSerializer(new ZNRecordSerializer());
-          HelixDataAccessor accessor =
-              new ZKHelixDataAccessor(helixClusterName, new ZkBaseDataAccessor<ZNRecord>(zkClient));
-          PropertyKey.Builder keyBuilder = accessor.keyBuilder();
-          accessor.setProperty(keyBuilder.stateModelDef(segmentStateModelName), newStateModelDef);
-          LOGGER.info("Completed updating statemodel {}", segmentStateModelName);
-          zkClient.close();
-        }
-      }
+      LOGGER.info("Helix cluster: {} already exists", helixClusterName);
       return;
     }
-
-    LOGGER.info("Creating a new cluster, as the helix cluster : " + helixClusterName
-        + " was not found ********************************************* ");
+    LOGGER.info("Creating a new Helix cluster: {}", helixClusterName);
     admin.addCluster(helixClusterName, false);
+    LOGGER.info("New Cluster: {} created.", helixClusterName);
+  }
+
+  private static HelixManager startHelixControllerInStandadloneMode(String helixClusterName, String zkUrl,
+      String pinotControllerInstanceId) {
+    LOGGER.info("Starting Helix Standalone Controller ... ");
+    return HelixControllerMain.startHelixController(zkUrl, helixClusterName, pinotControllerInstanceId,
+        HelixControllerMain.STANDALONE);
+  }
+
+  /**
+   * Customizes existing Helix cluster to run Pinot components.
+   */
+  public static void setupPinotCluster(String helixClusterName, String zkPath, boolean isUpdateStateModel,
+      boolean enableBatchMessageMode) {
+    final HelixAdmin admin = new ZKHelixAdmin(zkPath);
+    if (!admin.getClusters().contains(helixClusterName)) {
+      LOGGER.error("Helix cluster: {} hasn't been set up", helixClusterName);
+      throw new RuntimeException();
+    }
+
+    // Ensure auto join.
+    ensureAutoJoin(helixClusterName, admin);
+
+    // Add segment state model definition if needed
+    addSegmentStateModelDefinitionIfNeeded(helixClusterName, admin, zkPath, isUpdateStateModel);
+
+    // Add broker resource online offline state model definition if needed
+    addBrokerResourceOnlineOfflineStateModelDefinitionIfNeeded(helixClusterName, admin);
```

Comment:
addBrokerResourceOnlineOfflineStateModelDefinitionIfNeeded() and createBrokerResourceIfNeeded() can be combined into one right? (Same as the leadControllerResource)

Reply:
Merged.